### PR TITLE
Make DMLC output deterministic

### PR DESCRIPTION
**Pull request by mandolaerik**
- Adapt to rename of state-change notifier
- Enable line directives by default
- Allow disabling gcc optimizations
- Use dicts and lists instead of sets to avoid indeterministic output
- Sort sets before using, to avoid indeterministic output
- Validate that two runs with different pythons give the same C output
- fix minor typos
- Use ordinary dictionaries

### CODE REVIEW

**Comment by syssimics** [comment]
Verification [#10994099](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=10994099): fail

**Comment by mandolaerik** [comment]
All cases of inteterminisms were caused by set iteration order being undefined. This did not only cause differences between pypy and mini-python; it also caused differences between mini-python based runs.

Maintaining determinism adds a maintenance burden: if you use sets carelessly, this will not be captured until 226, which normally happens after merge, in Integration builds. We also need to define some guidelines for when sets can be used. The PR's current convention is roughly to avoid using collections with undefined iteration order *across functions*. This means:
* a set can be used within a function, but if the function constructs some list or dictionary based on it, then the set's elements must be sorted first
* If sharing data across functions which naturally would have a set datatype (no duplicate elements, order insignificant), then two options:
  * use list if possible (i.e. if fast `in` is not needed, and the collection is final)
  * otherwise, use a dict with None values, e.g. 
  * If nobody iterates the collection, then it's also possible to represent as a set, but this is brittle and should be a avoided
* Currently I made an exception for global reference counts, which are typically iterated once only. I found it more readable to represent this as a set, and sort it on iteration.
* Be generous about applying redundant sorting (not just deterministic order based on order of appearance in input) when this is cheap and potentially improves debugging experience, e.g. a centrally used collection like Trait.ancestors or when outputting a list of declarations/calls/etc. in C code, like vtable initialization calls.

These conventions are up for discussion, and the conventions changed over time so I haven't followed them strictly in this PR.

**Comment by syssimics** [comment]
Verification [#10994184](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=10994184): fail

**Comment by syssimics** [comment]
Verification [#10995016](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=10995016): pass

File: py/dml/traits.py [hunk 1]
```diff
@@ -444,7 +445,7 @@ def sort_method_implementations(implementations):
                             impl.name)
         rank_to_method[impl.rank] = impl
 
-    ranks = set(rank_to_method)
+    ranks = {r: None for r in rank_to_method}
```
Comment (JonatanWaern) [inline py/dml/traits.py:448]: For other reading this:
As of python 3.7 dictionaries retain insertion order for keys _as part of the official spec_, so this is deterministic if rank_to_method order is deterministic

**Review by JonatanWaern — APPROVED** [review]

**Comment by mandolaerik** [comment]
Some commits in this series are related to https://github.com/intel-innersource/applications.simulators.simics.simics-base/pull/3102

We could in principle dissociate test running parts of this PR into a separate  commit, but deterministic output is closely related to what t226 does so felt natural to combine them

**Comment by mandolaerik** [comment]
[V 3](https://teamcity01-ir.devtools.intel.com/viewQueued.html?itemId=11015149)

**Comment by awmoore-intel** [comment]
Will this result in output .c/.h files to be exactly the same, allowing users to use ccache in direct mode? https://ccache.dev/manual/latest.html#_how_ccache_works

**Comment by syssimics** [comment]
Verification [#11015149](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11015149): fail

**Comment by mandolaerik** [comment]
> Will this result in output .c/.h files to be exactly the same, allowing users to use ccache in direct mode? https://ccache.dev/manual/latest.html#_how_ccache_works

Yes

File: py/dml/set_test.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        self.assertFalse('bar' in x)
+        self.assertTrue(o1 in x)
+        self.assertFalse(o2 in x)
+
+    def test_eq(self):
+        x = Set(range(10))
+        self.assertEqual(x, Set(range(10)))
+        self.assertNotEqual(x, Set(range(11)))
+        self.assertEqual(x, set(range(10)))
+        assert set(range(10)) == {e: e for e in range(10)}.keys()
```
Comment (JonatanWaern) [inline py/dml/set_test.py:24]: this plain assert sticks out, why is this like this?

**Review by JonatanWaern — APPROVED** [review]
Other than a few lingering questions regarding when to use lists and when to use Sets, this seems fine to me

File: py/dml/set_test.py (hunk 2 above)
Comment (mandolaerik) [reply py/dml/set_test.py:24]: The class under test is `Set`, and this line doesn't test `Set` but is rather an executable comment that confirms that ordinary sets work the way that the previous tests expect Set to work (to me it's far from obvious that set and dict.keys() are expected to compare equal)

**Comment by syssimics** [comment]
Verification [#11020120](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11020120): fail

**Pull request merged**